### ISSUE TITLE
Feature request : Update backup

### ISSUE THREAD

**Issue by Zixim** [alekdavis/PlexBackup#6]
First of all : this might very well be the most elegantly written powershell script I've ever come across, personally or work-related.
Huge thumbs-up to @alekdavis !!

Being only human, we're never satisfied 🤦‍♂️ 
I took a backup yesterday, it took a looong time : 
```
Script ran for (hr:min:sec.msec):
  02:54:49.962
```
The `BackupRootDir` was set to a local drive, so that wasn't the cause of the duration.
The size of the Plex folders is at cause.

TL;DR : would you consider adding an option to make a differential or incremental backup? 
Perhaps the `-update` parameter to `Compress-Archive` might be useful ?

**Comment by alekdavis** [comment]
Thank you @Zixim. This was my first PowerShell script, so I gave it my best effort. :-)

When I had my Plex app data on a NAS share, backup took me a few hours. Once I moved it to a local drive, it is taking about 20-30 minutes. But I suspect that a bigger database would require more time.

I need to think about the incremental backup feature. Assuming that Compress-Archive can do incremental update, and I think it can, it's probably possible. The only questions would be: will it work faster and when will I have time. :-) My original thought was that for incremental backups, the Robocopy backup mode would work best, but I can give it a try in the compression mode. Don't promise though, but will give it a serious thought at least. 

Thanks for the feedback.

**Comment by Zixim** [comment]
Here's the stats of the data to be backed up, which took 2h:54m 
149k files for 14.8gb : 
```
Get-ChildItem $env:LOCALAPPDATA'\Plex Media Server' -recurse | Measure-Object -property length -sum


Count    : 149409
Average  :
Sum      : 14835940536
Maximum  :
Minimum  :
Property : length
```

**Comment by alekdavis** [comment]
I'll compare it with mine when I get to my home PC, but it looks close to mine (based on the sizes of the compressed files). Keep in mind that the backup ignores a few folders, although, it should not make a huge difference. Btw, when I first did it manually (before I had the script), it took me about 5 hours. :-)

**Comment by alekdavis** [comment]
Mine:
```
Count    : 108465
Average  : 
Sum      : 9029410238
Maximum  : 
Minimum  : 
Property : length
```

**Comment by alekdavis** [comment]
And last backup took 40 minutes.

**Comment by alekdavis** [comment]
Hey @Zixim, try the latest version; I added the -Update switch for incremental updates. In the test (not real Plex appdata, just my test folders), it does not really improve performance time much (if any).

**Comment by alekdavis** [comment]
Also, the update will add new files/folders to the ZIP files, but it will not remove the files/folders that were removed from the source since the last backup (that's how` Compress-Archive -Update` works).

**Comment by Zixim** [comment]
I'm running an update backup right now.

1st thing that strikes me, today's update-backup has nixed the log of the existing backup. From biological/unreliable memory : the previous took 1h40m. 

2nd is that, as you say, update isn't any quicker than normal backup.

Will try Robocopy & Robocopu-Update soon.

**Comment by alekdavis** [comment]
If you are asking about the script deleting the old log file, you can run it with the `-LogAppend` switch and the new log messages will be added to the old log file.

I suspected that differential backup would not be faster, since the archiver would need to first read the existing file, find the missing data, and then recompress the file. No idea how they do it internally, but given that Compress-Archive can't even handle file paths longer than 160 characters, I would not bet on it being in any way efficient. Robocopy, on the other hand, should be significantly faster.

**Comment by alekdavis** [comment]
Well, the update totally does not work for me. When I run the update, Compress-Archive just kills my USB flash drive hosting the backup folder (in the middle of archiving the Cache folder). No idea why. Tried multiple times and no luck. The normal backup works absolutely fine, but the normal backup archives to a temp folder first, and the update archives directly to the backup folder (I was thinking about moving it to the temp folder first, but then, in addition to extra code complexity, it would be too much overhead, i.e. time to takes to copy back and forth, that it'd offer little or no benefit or possibly be even worse). Anyway, I'll leave the update option as an experimental feature, but if you are mostly concerned with the backup time and want to use incremental updates, just use the Robocopy option.

**Comment by Zixim** [comment]
fwiw... robocopy differential is worth it for me.
My main goal is to use syncthing for offsite duplication of the backup.
After lots of testing, Robo + Update is perfect for this.

**Comment by alekdavis** [comment]
Cool. I'm going to close this issue then. Let me know if you run into any issues.

**Comment by alekdavis** [comment]
I reverted back to not supporting the Update option because at least in theory it should be slower than original compression (because it first needs to decompress the existing archive and then re-compress it back). At least, this is what the 7-zip documentation says, and I'm sure Windows compression works the same way.  I added support for 7-zip which seems to work a lot faster and produce smaller files.

**Comment by Zixim** [comment]
but update remains for Robocopy ?

**Comment by alekdavis** [comment]
Use the `-Continue` flag.
